File: src/shepherd/feed.py
from dataclasses import dataclass
from typing import NamedTuple
from enum import Enum, auto


from shepherd.ids import TagId, ItemId


class Response(Enum):
    """An enum indicating a Sheep's response to a Feed Item"""

    POSITIVE = auto()
    NEUTRAL = auto()
    NEGATIVE = auto()


@dataclass(frozen=True)
class Responses:
    """
    The responses returned from a Sheep after evaluating a Feed

    These must be in the same order as those provided in the original Feed,
    to prevent an incorrect assignment of Sheep sentiment toward the Feed's
    Items
    """

    items: list[Response]


@dataclass(frozen=True)
class Item:
    """
    An item within the simulation

    This can represent anything. All it has is an ID and associated tags
    """

    id: ItemId
    tags: list[TagId]


@dataclass(frozen=True)
class Feed:
    """The feed returned from a Shepherd, prepared for specific Sheep"""

    items: list[ItemId]

File: lib/src/feed.rs
use serde::{Deserialize, Serialize};

use crate::ids::ItemId;

/// An enum indicating a Sheep's response to a [`Feed`] item
#[derive(Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
pub enum Response {
    /// A positive response to a feed item
    Positive,

    /// A neutral response to a feed item
    Neutral,

    /// A negative response to a feed item
    Negative,
}

/// The [`Response`]s returned from a Sheep after evaluating a [`Feed`]
///
/// The first two values are self-explanatory, the third is a count of how many
/// hops are required to get from the Sheep to the feed item (if it is
/// reachable)
#[derive(Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
pub struct Responses(pub Vec<(ItemId, Response, Option<u32>)>);

/// The feed returned from a [`Shepherd`], prepared for a specific Sheep
#[derive(Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
pub struct Feed(pub Vec<ItemId>);
